File: upwind/julia/upwind.jl
################################################################################
# Upwind advection
# 
# Original code from Christopher Blanton (cjb47@psu.edu)
# Modified by Alex Pletzer (alexander@gokliya.net)

using LinearAlgebra 

include("saveVTK.jl")

mutable struct Upwind

    # Number of space dimensions
    ndims::Int

    # Constant advection velocity
    velocity::Array

    # Domain length
    lengths::Array

    # Number of cells along each direction
    numCells::Array

    #
    # Internal variables (derived from the above)
    #

    # Grid size along each direction
    deltas::Array
    # Direction of upwind
    upDirection::Array
    # Total number of cells
    ntot::Int
    # Product of the dimensions, used to go from flat index to index set
    dimProd::Array
    # Field
    f::Array

    function Upwind(ndims::Int, velocity::Array, lengths::Array, numCells::Array)
        """
        Internal constructor
        @param ndims number of space dimensions
        @param velocity velocity array of size ndims
        @param lengths domain lengths array of size ndims
        @param numCells number of cells, array of size ndims
        """
        this = new(ndims, velocity, lengths, numCells)

        this.deltas = zeros(Float64, ndims::Integer)

        this.upDirection = zeros(Integer, ndims::Integer)
        this.ntot = 1
        for i in 1:ndims
            this.upDirection[i] = -1
            if this.velocity[i] < 0.0
                this.upDirection[i] = +1
            end
            this.deltas[i] = lengths[i]/numCells[i]
            this.ntot *= numCells[i]
        end
        this.dimProd = ones(Integer, ndims::Integer)
        for i = 2:ndims
            this.dimProd[i] = this.dimProd[i-1] * this.numCells[i-1]
        end

        # 
        # Set the initial field
        #
        this.f = zeros(this.ntot)
        this.f[1] = 1.0

        return this
    end
end

function advect!(this::Upwind, deltaTime::Float64)
    """
    Advance the solution by one time step
    @param this instance
    @param deltaTime time increment
    """
    oldF = deepcopy(this.f)
    for j = 1:this.ndims
        c = deltaTime * this.velocity[j] * this.upDirection[j] / this.deltas[j]
        for i = 1:this.ntot
            inds = getIndexSet(up, i)
            oldIndex = inds[j]
            #periodic BCs
            inds[j] += this.upDirection[j]
            inds[j] = mod1(inds[j], this.numCells[j])
            upI = getFlatIndex(up, inds)
            this.f[i] -= c * (oldF[upI] - oldF[i])
            inds[j] = oldIndex
        end
    end 
end

function getFlatIndex(this::Upwind, inds::Array)
    """
    Get the flat index from an index set
    @param this instance
    @param inds index set, eg [i, j, k]
    @return integer index of field array
    """
    return dot(this.dimProd, inds .- 1) + 1
end
  
function getIndexSet(this::Upwind, flatIndex::Integer)
    """
    Get the index set of a falt index
    @param this instance
    @param flatIndex flat index 
    @return index set, eg [i, j, k]
    """
    res = zeros(Integer, this.ndims)
    for i = 1:this.ndims
        res[i] = mod(div((flatIndex - 1), this.dimProd[i]), this.numCells[i]) + 1
    end
    return res
end

function save2VTK(this::Upwind, fname::AbstractString)
    """
    Save data in VTK file
    @param this instance
    @param fname file name
    """
    xAxis = linspace(0.0, this.lengths[1], this.numCells[1] + 1)
    yAxis = linspace(0.0, this.lengths[2], this.numCells[2] + 1)
    zAxis = linspace(0.0, this.lengths[3], this.numCells[3] + 1)
    saveVTK.rectilinear(fname, xAxis, yAxis, zAxis, this.f)

end

# Beginning of main program

ndims = 3

# Parse command line arguments
if length(ARGS) < 1
    println("Must specify number of cells in each direction.")
    println("Usage: ", basename(Base.source_path()), " numCells [numTimeSteps]")
    exit(1)
end
ncells = parse(Int, ARGS[1])

numTimeSteps = 100
if length(ARGS) > 1
    numTimeSteps = parse(Int, ARGS[2])
end

# Same resolution in each direction.
numCells = [ncells,ncells,ncells]
println("number of cells: ", numCells)
println("number of time steps: ", numTimeSteps)


# Initalizing vector velocity and lengths
velocity = ones(Float64, ndims)
lengths   = ones(Float64, ndims)


# Compute time step
courant = 0.1
dt = Inf
for i = 1:ndims
    dx = lengths[i]/float(numCells[i])
    global dt = min((courant*dx)/velocity[i], dt)
end

# Create Upwind instance
up = Upwind(ndims,velocity,lengths,numCells)

# Advect
for i = 1:numTimeSteps
   advect!(up, dt)
end 

# Do the checksum
println("check sum: ", sum(up.f))

if length(ARGS) > 2 && ARGS[3] == "vtk"
   save2VTK(up, "upwind.vtk")
end

File: upwind/python/upwind.py
#/usr/bin/env python

from __future__ import print_function
import numpy
import copy
import saveVTK

class Upwind: 

  def __init__(self, velocity, lengths, numCells):
    self.numCells = numCells
    self.ndims = len(velocity)
    self.deltas = numpy.zeros( (self.ndims,), numpy.float64 )
    self.upDirection = numpy.zeros( (self.ndims,), numpy.int32 )
    self.v = velocity
    self.lengths = lengths
    self.ntot = 1
    self.offsetMat = numpy.identity(self.ndims, numpy.int32)
    self.numCellsExt = numpy.outer(self.numCells, numpy.ones((self.ndims,), numpy.int32))
    for j in range(self.ndims):
      self.upDirection[j] = -1
      if velocity[j] < 0.: self.upDirection[j] = +1
      self.deltas[j] = lengths[j] / numCells[j]
      self.offsetMat[j, j] = self.upDirection[j]
      self.ntot *= numCells[j]

    self.dimProd = numpy.ones( (self.ndims,), numpy.int32 )
    for i in range(self.ndims - 2, -1, -1):
        # last index varies fastest
        self.dimProd[i] =  self.dimProd[i + 1] * self.numCells[i + 1]

    self.coeff = self.v * self.upDirection / self.deltas

    # initializing the field
    self.f = numpy.zeros( (self.ntot,), numpy.float64 )
    # initialize lower corner to one
    self.f[0] = 1

    # array of index sets for each cell
    self.inds = numpy.zeros( (self.ndims, self.ntot), numpy.int32 )
    for j in range(self.ndims):
      self.inds[j, :] = numpy.arange(self.ntot)
      self.inds[j, :] //= self.dimProd[j]
      self.inds[j, :] %= self.numCells[j]

  def advect(self, deltaTime):

    # copy 
    oldF = self.f.copy()

    indsUp = self.inds.copy()

    # update the field in each spatial direction
    for j in range(self.ndims):

      # apply offset
      indsUp[j, :] += self.upDirection[j]
      indsUp[j, :] %= self.numCells[j]

      # compute flat indices corresponding to the offset index sets
      flatIndsUp = numpy.dot(self.dimProd, indsUp)

      # update
      self.f -= (deltaTime * self.coeff[j]) * (oldF[flatIndsUp] - oldF)

      # reset
      indsUp[j, :] = self.inds[j, :]

  def saveVTK(self, fname):
    xAxis = [0.0]
    yAxis = [0.0]
    zAxis = [0.0]
    if self.ndims > 2:
      xAxis = [0.0 + self.deltas[2] * i for i in range(self.numCells[2] + 1)]
    if self.ndims > 1:
      yAxis = [0.0 + self.deltas[1] * i for i in range(self.numCells[1] + 1)]
    if self.ndims > 0:
      zAxis = [0.0 + self.deltas[0] * i for i in range(self.numCells[0] + 1)]
    saveVTK.rectilinear(fname, xAxis, yAxis, zAxis, self.f.reshape(self.numCells))

  def checksum(self):
    return numpy.sum(self.f)

  def printOut(self):
    for i in range(len(self.f)):
      print(i, ' ', self.f[i])

#################################################################################################
def main():
  import sys

  if len(sys.argv) <= 1:
    print("must specify number of cells in each direction.")
    return sys.exit(1)

  ndims = 3
  numCells = [int(sys.argv[1])] * 3
  print("number of cells: ", numCells)

  numTimeSteps = 100
  if len(sys.argv) > 2:
    numTimeSteps = int(sys.argv[2])
  print("number of time steps: ", numTimeSteps)

  doVtk = False
  if len(sys.argv) > 3 and sys.argv[3] == 'vtk':
    doVtk = True

  velocity = numpy.ones( (ndims,), numpy.float64 )
  lengths = numpy.ones( (ndims,), numpy.float64 )

  # compute dt 
  courant = 0.1
  dt = float('inf')
  for j in range(ndims):
    dx = lengths[j]/ float(numCells[j])
    dt = min(courant * dx / velocity[j], dt)

  up = Upwind(velocity, lengths, numCells)
  for i in range(numTimeSteps):
    up.advect(dt)

  print("check sum: ", up.checksum())
  if doVtk:
    up.saveVTK("up.vtk")

if __name__ == '__main__': main()
